crucible/test/helpers/Main.hs

```diff
@@ -1,20 +1,84 @@
 {-# LANGUAGE AllowAmbiguousTypes #-}
 module Main where
 
+import Control.Lens ((^.))
 import Data.List (isInfixOf)
+import Data.Maybe (fromMaybe)
 
 import Test.Hspec
 import Test.Tasty
 import Test.Tasty.Hspec (testSpec)
+import qualified Test.Tasty.HUnit as TTH
 
+import Data.Parameterized.Nonce (newIONonceGenerator)
+import Data.Parameterized.Some (Some(Some))
+import Lang.Crucible.Backend (SomeBackend(..))
+import qualified Lang.Crucible.Backend as LCB
+import Lang.Crucible.Backend.Simple (newSimpleBackend)
 import Lang.Crucible.Panic
+import Lang.Crucible.Simulator.SimError (SimErrorReason(GenericSimError))
+import What4.Expr (EmptyExprBuilderState(..))
+import What4.Expr.Builder (newExprBuilder)
+import qualified What4.Interface as W4I
+import What4.FloatMode (FloatModeRepr(FloatIEEERepr))
+import qualified What4.LabeledPred as W4L
+import qualified What4.ProgramLoc as W4P
 
 import qualified Panic as P
 
 main :: IO ()
 main =
   defaultMain =<< panicTests
 
+mkBackend :: IO (Some SomeBackend)
+mkBackend = do
+  Some nonceGen <- newIONonceGenerator
+  sym <- newExprBuilder FloatIEEERepr EmptyExprBuilderState nonceGen
+  Some . SomeBackend <$> newSimpleBackend sym
+
+assumePred :: LCB.IsSymBackend sym bak => bak -> W4I.Pred sym -> IO ()
+assumePred bak p =
+  LCB.addAssumption bak (LCB.GenericAssumption W4P.initializationLoc "" p)
+
+backendTests :: TestTree
+backendTests =
+  testGroup
+  "Backend"
+  [ TTH.testCase "push/pop nothing" $ do
+      Some (SomeBackend bak) <- mkBackend
+      asmps <- LCB.popAssumptionFrame bak =<< LCB.pushAssumptionFrame bak
+      p <- LCB.assumptionsPred (LCB.backendGetSym bak) asmps
+      Just True TTH.@=? W4I.asConstantPred p
+  , TTH.testCase "push/pop assumptions" $ do
+      Some (SomeBackend bak) <- mkBackend
+      let sym = LCB.backendGetSym bak
+      a <- W4I.freshConstant sym (W4I.safeSymbol "a") W4I.BaseBoolRepr
+      assumePred bak a
+      frm <- LCB.pushAssumptionFrame bak
+      b <- W4I.freshConstant sym (W4I.safeSymbol "b") W4I.BaseBoolRepr
+      assumePred bak b
+      p <- LCB.assumptionsPred sym =<< LCB.popAssumptionFrame bak frm
+      pEqB <- W4I.eqPred sym p b
+      Just True TTH.@=? W4I.asConstantPred pEqB
+  , TTH.testCase "push/pop obligations" $ do
+      Some (SomeBackend bak) <- mkBackend
+      let sym = LCB.backendGetSym bak
+      a <- W4I.freshConstant sym (W4I.safeSymbol "a") W4I.BaseBoolRepr
+      assumePred bak a
+      frm <- LCB.pushAssumptionFrame bak
+      b <- W4I.freshConstant sym (W4I.safeSymbol "b") W4I.BaseBoolRepr
+      assumePred bak b
+      c <- W4I.freshConstant sym (W4I.safeSymbol "c") W4I.BaseBoolRepr
+      LCB.assert bak c (GenericSimError "")
```

Comment:
Worth giving a more descriptive string than "" here?

Reply:
I'm not exactly sure what to put considering that this is asserting a fresh symbolic constant. I could put something like `"c"` or `"test"`, but I'm not sure how much of an improvement that would be.

Reply:
Added `"asserting c"`